elife/multi-process-test.sls

```diff
@@ -0,0 +1,48 @@
+
+# test that N processes are brought using a single service state
+# each process simply writes it's identifier and the time to the syslog
+
+{% set process = "test" %}
+{% set num_processes = 10 %}
+
+/opt/{{ process }}.py:
+    file.managed:
+        - source: salt://elife/scripts/test.py
+
+{{ process }}:
+    file.managed:
+        - name: /lib/systemd/system/{{ process }}@.service
+        - source: salt://elife/config/lib-systemd-system-test.service
+        - template: jinja
+        - context:
+            process: {{ process }}
+        - require:
+            - /opt/{{ process }}.py
+
+    # neccessary because salt mangles "service@{x..y}" to "service@\x7bx..y\x7d.service" 
+    # systemd will see that as just a single instantiated service
+    cmd.run:
+        - name: |
+            set -e
+            systemctl disable {{ process }}@{0..99} # disable all instances (up to 99). implicit reload
```

Comment:
could it just `stop` and `start` rather than `disable` and `enable`? The difference is whether it affects what happens at boot and so the time interval in which `cmd.run` is executed it doesn't make a difference.

Reply:
If there are many processes and `disable` waits for them, there may be a performance problem. In the Upstart script we send stopping signals for all of them asynchronously (no waiting) and then wait for termination of everything as a second step.

Reply:
Maybe this doesn't actually stop and start them now, which would make sense.

Reply:
this example isn't quite finished yet. enable and disable for template services determines how many are started when the controller becomes available. 

if we enable 5 processes and restart the controller, then 5 processes will be started.

if we change that to 3 processes and restart the controller, we want just 3 processes running but we still have 2 others that weren't disabled.

you can target processes individually with `service@n`, and we do want some feedback in the formula if the service failed to start, so we could have a `for` loop of `service.running` states in the formula.

Reply:
> If there are many processes and disable waits for them

it doesn't. `enable` and `disable` are a bit like `a2enmod` and `a2dismod` - they're just wrappers around creating symlinks.

when the controller state changes these processes are brought up or stopped asynchronously